### PR TITLE
[clusteragent/autoscaling] Add telemetry metrics for autoscaling controller

### PR DESCRIPTION
**Pull request by jennchenn**
### What does this PR do?

Add the following telemetry metrics around the autoscaling controller:
* `autoscaling.queue_*`
* `autoscaling.status_conditions`
* `workload_autoscaling.horizontal_scaling`
* `workload_autoscaling.vertical_scaling`

### Motivation

Improve monitoring and debugging.

### Describe how to test/QA your changes

Deploy Cluster Agent with autoscaling controller activated, `autoscaling.*` metrics should be available on Prom endpoint (`localhost:5000/metrics`).
With Agent >= 7.57, metrics should also be scraped by DCA integration

### CODE REVIEW

File: pkg/clusteragent/autoscaling/workload/controller_horizontal.go [hunk 1]
```diff
@@ -133,13 +134,15 @@ func (hr *horizontalController) performScaling(ctx context.Context, podAutoscale
 	if err != nil {
 		err = fmt.Errorf("failed to scale target: %s/%s to %d replicas, err: %w", scale.Namespace, scale.Name, horizontalAction.ToReplicas, err)
 		hr.eventRecorder.Event(podAutoscaler, corev1.EventTypeWarning, model.FailedScaleEventReason, err.Error())
+		telemetryHorizonalScale.Inc(scale.Namespace, scale.Name, strconv.Itoa(int(horizontalAction.ToReplicas)), strconv.Itoa(int(horizontalAction.FromReplicas)), strconv.Itoa(int(*horizontalAction.RecommendedReplicas)), "true")
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/controller_horizontal.go:137]: For values, it's easier to expose a `Gauge` with the `ToReplicas` value (basically seeing history). We can counters for number of success/failures, not tracking the `To/From` (not needed). It's interesting to have the target `namespace` and `name` but we also want the `DatadogPodAutoscaler` namespace and name (`namespace` is the same in both cases and could be omitted)

File: pkg/clusteragent/autoscaling/controller.go [hunk 2]
```diff
@@ -54,8 +54,14 @@ func NewController(
 		Client:    client,
 		Lister:    mainInformer.Lister(),
 		synced:    mainInformer.Informer().HasSynced,
-		Workqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter()),
-		IsLeader:  isLeader,
+		Workqueue: workqueue.NewRateLimitingQueueWithConfig(
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/controller.go:57]: Controller is generic (even though in `autoscaling` currently, we should probably pass the workqueue as parameter.

File: pkg/clusteragent/autoscaling/workload/model/pod_autoscaler.go [hunk 3]
```diff
@@ -499,6 +508,11 @@ func (p *PodAutoscalerInternal) BuildStatus(currentTime metav1.Time, currentStat
 	}
 	status.Conditions = append(status.Conditions, newCondition(rolloutStatus, verticalReason, currentTime, datadoghq.DatadogPodAutoscalerVerticalAbleToApply, existingConditions))
 
+	for _, condition := range status.Conditions {
+		if condition.Status == corev1.ConditionTrue {
+			autoscalingStatusConditions.Inc(string(condition.Type), condition.Reason, condition.Message)
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/model/pod_autoscaler.go:513]: It's probably more a 0-1 gauge (1 = True, 0 = everything else). The telemetry should probably be in the controller code and not in the `model` package.

File: pkg/clusteragent/autoscaling/workload/model/pod_autoscaler.go [hunk 4]
```diff
@@ -30,6 +31,14 @@ const (
 	statusRetainedActions = 5
 )
 
+var autoscalingStatusConditions = telemetry.NewCounterWithOpts(
+	"autoscaling",
+	"status_condition",
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/model/pod_autoscaler.go:36]: The generated metric name will not be very clear. We would expect something like `datadog.cluster_agent.autoscaling.autoscaler_conditions`

File: pkg/clusteragent/autoscaling/workload/model/pod_autoscaler.go [hunk 5]
```diff
@@ -30,6 +31,14 @@ const (
...
+	[]string{"type", "reason", "message"},
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/model/pod_autoscaler.go:37]: You always need autoscaler name and namespace.

File: pkg/clusteragent/autoscaling/workload/telemetry.go [hunk 6]
```diff
@@ -17,11 +22,39 @@ const (
...
 var commonOpts = telemetry.Options{NoDoubleUnderscoreSep: true}
 
-// rolloutTriggered tracks the number of patch requests sent by the patcher to the kubernetes api server
-var rolloutTriggered = telemetry.NewCounterWithOpts(
-	subsystem,
-	"rollout_triggered",
-	[]string{"owner_kind", "owner_name", "namespace", "status"},
-	"Tracks the number of patch requests sent by the patcher to the kubernetes api server",
-	commonOpts,
+func resourceListToString(m map[corev1.ResourceName]resource.Quantity) string {
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/telemetry.go:25]: nit:  `func` go after `var` block

File: pkg/clusteragent/autoscaling/workload/controller_vertical.go [hunk 7]
```diff
@@ -69,9 +69,32 @@ func (u *verticalController) sync(ctx context.Context, podAutoscaler *datadoghq.
 	targetGVK, err := autoscalerInternal.TargetGVK()
 	if err != nil {
 		autoscalerInternal.SetError(err)
+		for _, resource := range scalingValues.Vertical.ContainerResources {
+			telemetryVerticalScale.Inc(
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/controller_vertical.go:73]: Same here, we want a `Gauge` with target value being the the limit or request value.

File: pkg/clusteragent/autoscaling/workload/controller_horizontal.go [hunk 8]
```diff
@@ -133,13 +134,15 @@ func (hr *horizontalController) performScaling(ctx context.Context, podAutoscale
...
 		autoscalerInternal.UpdateFromHorizontalAction(nil, err)
 		return autoscaling.Requeue, err
 	}
 
 	log.Debugf("Scaled target: %s/%s from %d replicas to %d replicas", scale.Namespace, scale.Name, horizontalAction.FromReplicas, horizontalAction.ToReplicas)
 	autoscalerInternal.UpdateFromHorizontalAction(horizontalAction, nil)
 	hr.eventRecorder.Eventf(podAutoscaler, corev1.EventTypeNormal, model.SuccessfulScaleEventReason, "Scaled target: %s/%s from %d replicas to %d replicas", scale.Namespace, scale.Name, horizontalAction.FromReplicas, horizontalAction.ToReplicas)
+	telemetryHorizonalScale.Inc(scale.Namespace, scale.Name, strconv.Itoa(int(horizontalAction.ToReplicas)), strconv.Itoa(int(horizontalAction.FromReplicas)), strconv.Itoa(int(*horizontalAction.RecommendedReplicas)), "false")
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/controller_horizontal.go:145]: Currently you're tracking _applied_ recommendations, but not received recommendations. We're interested in both

File: pkg/clusteragent/autoscaling/workload/telemetry.go [hunk 9]
```diff
@@ -17,11 +22,39 @@ const (
...
+	rolloutTriggered = telemetry.NewCounterWithOpts(
+		subsystem,
+		"rollout_triggered",
+		[]string{"owner_kind", "owner_name", "namespace", "status"},
+		"Tracks the number of patch requests sent by the patcher to the kubernetes api server",
+		commonOpts,
+	)
+
+	// telemetryHorizonalScale tracks the horizontal scaling recommendation values
+	telemetryHorizonalScale = telemetry.NewCounterWithOpts(
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/telemetry.go:44]: Note that you need to differentiate metrics that can only be emitted by the leader and add the `le.JoinLeaderLabel` (you can check how it works in the externalmetrics package).

**Review by vboulineau — APPROVED** [review]

**Review by ogaca-dd — APPROVED** [review]
LGTM for files owned by ASC

**Pull request merged**